### PR TITLE
Fast path Text from native code

### PR DESCRIPTION
**Pull request by jnvbui**
Added fast path rendering for textblocks when possible.

Currently textblocks are created using Inlines in react-native, even when the textblock doesn't have any nested children, which negatively affects performance in UWP. More documentation [here](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/text-block#performance-considerations).

Modified from https://github.com/microsoft/react-native-windows/pull/2777
Avoided editing from Text.js so don't need to fork code from react-native. Instead edited RawTextViewManager and TextViewManager.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/2798)

### CODE REVIEW

**Comment by jnvbui** [comment]
Fixes #2620

File: vnext/ReactUWP/Views/RawTextViewManager.cpp [hunk 1]
```diff
@@ -12,6 +12,8 @@
 #include <winrt/Windows.Foundation.h>
 #include <winrt/Windows.UI.Xaml.Controls.h>
 #include <winrt/Windows.UI.Xaml.Documents.h>
+#include <ReactUWP\Base\UwpReactInstance.h>
```
Comment (stecrain) [inline vnext/ReactUWP/Views/RawTextViewManager.cpp]: Do you need to include this?

File: vnext/ReactUWP/Views/TextViewManager.cpp [hunk 2]
```diff
@@ -22,11 +22,37 @@ namespace uwp {
...
+        auto textBlock = this->GetView().as<winrt::TextBlock>();
+        textBlock.Text(run.Text());
+        return;
+      }
+    }
+    else if (index == 1 && m_firstChildNode != nullptr) {
+      auto textBlock = this->GetView().as<winrt::TextBlock>();
+      textBlock.ClearValue(winrt::TextBlock::TextProperty());
+      Super::AddView(*m_firstChildNode, 0);
+      m_firstChildNode = nullptr;
```
Comment (stecrain) [inline vnext/ReactUWP/Views/TextViewManager.cpp]: I think you should also null this out in 
```
void ShadowNodeBase::removeAllChildren() {
  GetViewManager()->RemoveAllChildren(GetView());
}
```

override removeAllChildren so you can null it out.

**Comment by stecrain** [comment]
@kmelmon Fast path text with a native only solution

File: vnext/ReactUWP/Views/RawTextViewManager.cpp [hunk 3]
```diff
@@ -51,6 +52,19 @@ void RawTextViewManager::UpdateProperties(
 
     if (propertyName == "text") {
       run.Text(asHstring(propertyValue));
+      if (nodeToUpdate->GetParent() != -1) {
+        auto parent = this->m_wkReactInstance.lock()
+                        ->NativeUIManager()
+                        ->getHost()
+                        ->FindShadowNodeForTag(nodeToUpdate->GetParent());
+        if (parent->m_children.size() == 1) {
+          auto view = static_cast<ShadowNodeBase &>(*parent).GetView();
```
Comment (stecrain) [inline vnext/ReactUWP/Views/RawTextViewManager.cpp]: parent->GetView()
?

File: vnext/ReactUWP/Views/TextViewManager.cpp [hunk 4]
```diff
@@ -22,11 +22,41 @@ namespace uwp {
...
+    m_firstChildNode = nullptr;
+  };
   bool ImplementsPadding() override {
     return true;
   }
+
+  void AddView(ShadowNode &child, int64_t index) override {
+    if (index == 0) {
+      auto run =
+          static_cast<ShadowNodeBase &>(child).GetView().try_as<winrt::Run>();
```
Comment (stecrain) [inline vnext/ReactUWP/Views/TextViewManager.cpp]: child->GetView()
?

Comment (jnvbui) [reply vnext/ReactUWP/Views/TextViewManager.cpp]: has to be ShadowNodeBase to call getView

File: vnext/ReactUWP/Views/TextViewManager.cpp [hunk 5]
```diff
@@ -22,11 +22,40 @@ namespace uwp {
...
+    } else if (index == 1 && m_firstChildNode != nullptr) {
+      auto textBlock = this->GetView().as<winrt::TextBlock>();
+      textBlock.ClearValue(winrt::TextBlock::TextProperty());
+      Super::AddView(*m_firstChildNode, 0);
+      m_firstChildNode = nullptr;
+    }
+    Super::AddView(child, index);
+  }
+
+  void removeAllChildren() {
```
Comment (kmelmon) [inline vnext/ReactUWP/Views/TextViewManager.cpp]: >removeAllChildren [](start = 7, length = 17)

There's also a RemoveChildAt() in ShadowNodeBase.  If this can remove child 0 then we have another case where we need to null out m_firstChildNode.

File: vnext/ReactUWP/Views/TextViewManager.cpp [hunk 6]
```diff
@@ -22,11 +22,40 @@ namespace uwp {
...
+    m_firstChildNode = nullptr;
+    Super::removeAllChildren();
+  }
```
Comment (kmelmon) [inline vnext/ReactUWP/Views/TextViewManager.cpp]: There's a non-trivial number of edge cases here.  Does RNTester exercise these?  If not please add test cases to RNTester.

Comment (stecrain) [reply vnext/ReactUWP/Views/TextViewManager.cpp]: @jnvbui The test cases we had in RNTester should be committed.
- Text moving from fast path to slow path
- Changing the Text in a single run (Why the run needs to be able to update the parent)

I think those are the main ones. but we should include the whole enumeration of them

Comment (jnvbui) [reply vnext/ReactUWP/Views/TextViewManager.cpp]: just added fast path text test cases to rntester, let me know if more are needed

Comment (kmelmon) [reply vnext/ReactUWP/Views/TextViewManager.cpp]: Thanks for adding tests!

---
In reply to: [305156938](https://github.com/microsoft/react-native-windows/pull/2798#discussion_r305156938) [](ancestors = 305156938)

Comment (jnvbui) [reply vnext/ReactUWP/Views/TextViewManager.cpp]: np! added one more test for moving from fast to slow path
